File: src/datatypes/note.ts
import { BaseEntity, EntityType, NotePriority } from './types';

export class Note implements BaseEntity {
    id!: string;
    user_id!: string;
    timestamp!: Date;
    content!: string;
    priority: NotePriority = NotePriority.MEDIUM;
    parent_id?: string;
    tags?: string[];
    embedding?: number[];
    metadata?: Record<string, any>;
    created_at!: Date;
    updated_at!: Date;

    constructor(data: Partial<Note>) {
        Object.assign(this, data);
    }
}

export class NoteReference {
    note_id!: string;
    reference_type!: string;
    reference_id!: string;
    timestamp?: Date;
    metadata?: Record<string, any>;
    created_at!: Date;

    constructor(data: Partial<NoteReference>) {
        Object.assign(this, data);
    }
}

export class KnownEntity implements BaseEntity {
    entity_id!: string;
    type!: EntityType;
    label!: string;
    embedding?: number[];
    metadata?: Record<string, any>;
    created_at!: Date;
    updated_at!: Date;

    constructor(data: Partial<KnownEntity>) {
        Object.assign(this, data);
    }
} 

File: src-tauri/src/datatypes/note.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use crate::datatypes::types::{EntityType, NotePriority, Metadata};

#[derive(Debug, Serialize, Deserialize)]
pub struct Note {
    pub id: String,
    pub user_id: String,
    pub timestamp: DateTime<Utc>,
    pub content: String,
    #[serde(default = "default_note_priority")]
    pub priority: NotePriority,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub parent_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tags: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub embedding: Option<Vec<f32>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<Metadata>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct NoteReference {
    pub note_id: String,
    pub reference_type: String,
    pub reference_id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub timestamp: Option<DateTime<Utc>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<Metadata>,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct KnownEntity {
    pub entity_id: String,
    pub entity_type: EntityType,
    pub label: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub embedding: Option<Vec<f32>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<Metadata>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

fn default_note_priority() -> NotePriority {
    NotePriority::Medium
}

impl Note {
    pub fn from_json(json: &str) -> Result<Self, serde_json::Error> {
        serde_json::from_str(json)
    }

    pub fn to_json(&self) -> Result<String, serde_json::Error> {
        serde_json::to_string(self)
    }
}

impl NoteReference {
    pub fn from_json(json: &str) -> Result<Self, serde_json::Error> {
        serde_json::from_str(json)
    }

    pub fn to_json(&self) -> Result<String, serde_json::Error> {
        serde_json::to_string(self)
    }
}

impl KnownEntity {
    pub fn from_json(json: &str) -> Result<Self, serde_json::Error> {
        serde_json::from_str(json)
    }

    pub fn to_json(&self) -> Result<String, serde_json::Error> {
        serde_json::to_string(self)
    }
} 